File: solution/147. Insertion Sort List/147.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        ListNode dummy = new ListNode();
        ListNode pre = dummy;
        while (head != null) {
            ListNode temp = head.next;
            if (head.val < pre.val) pre = dummy;
            while (pre.next != null && pre.next.val < head.val)
                pre = pre.next;
            head.next = pre.next;
            pre.next = head;
            head = temp;
        }
        return dummy.next;
    }
}

File: solution/147. Insertion Sort List/147.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def insertionSortList(self, head):
        dummy = ListNode(0)
        pre = dummy
        while head:
            temp = head.next
            if head.val < pre.val:
                pre = dummy
            while pre.next and pre.next.val < head.val:
                pre = pre.next
            head.next = pre.next
            pre.next = head
            head = temp
        return dummy.next
        
